src/uu/rm/BENCHMARKING.md

```diff
@@ -0,0 +1,52 @@
+# Benchmarking rm
+
+Run `cargo build --release` before benchmarking after you make a change!
+
+## Simple recursive rm
+
+- Get a large tree, for example linux kernel source tree.
+- We'll need to pass a `--prepare` argument, since `rm` deletes the dir each time.
+- Benchmark simple recursive rm with hyperfine: `hyperfine --prepare "cp -r tree tree-tmp" "target/release/coreutils rm -r tree-tmp"`.
+
+## Comparing with GNU rm
+
+Hyperfine accepts multiple commands to run and will compare them. To compare performance with GNU rm
+duplicate the string you passed to hyperfine but remove the `target/release/coreutils` bit from it.
+
+Example: `hyperfine --prepare "cp -r tree tree-tmp" "target/release/coreutils rm -rf tree-tmp"` becomes
+`hyperfine --prepare "cp -r tree tree-tmp" "target/release/coreutils rm -rf tree-tmp" "rm -rf tree-tmp"`
+(This assumes GNU rm is installed as `rm`)
+
+This can also be used to compare with version of rm built before your changes to ensure your change does not regress this.
+
+Here is a `bash` script for doing this comparison:
+
+```shell
+#!/bin/bash
+cargo build --no-default-features --features rm --release
+test_dir="$1"
+hyperfine --prepare "cp -r $test_dir tmp_d"  "rm -rf tmp_d" "target/release/coreutils rm -rf tmp_d"
+```
+
+## Checking system call count
+
+- Another thing to look at would be system calls count using strace (on linux) or equivalent on other operating systems.
+- Example: `strace -c target/release/coreutils rm -rf tree`
+
+## Cargo Flamegraph
+
+With Cargo Flamegraph you can easily make a flamegraph of `rm`:
+
+```shell
+cargo flamegraph --cmd coreutils -- rm [additional parameters]
+```
+
+However, if the `-r` option is given, the output becomes pretty much useless due to recursion. We can fix this by merging all the direct recursive calls with `uniq`, below is a `bash` script that does this.
+
+```shell
+#!/bin/bash
+cargo build --release --no-default-features --features rm
+perf record target/release/coreutils rm "$@"
```

Comment:
you might want to try https://github.com/mstange/samply

Reply:
That's interesting, just tried samply and `rm -rf` immediately enters into `std::sys::unix::fs::remove_dir_impl::remove_dir_all_recursive`, which is then 100% of the rest of the samples.

This is likely why `rm -rf` performance isn't improved by `jwalk` at all. We might see improvements if we were manually overriding the `remove_dir_all_recursive`, which appears to be causing [other issues](https://github.com/uutils/coreutils/issues/3722) with GNU compatibility.

Reply:
This test was again using the linux kernel with `rm -rf`

![image](https://github.com/uutils/coreutils/assets/47790940/eeb1c844-5019-4756-83d4-941e0cec82d3)